### PR TITLE
[ENH] fill_direction improve

### PR DESCRIPTION
**Pull request by samukweku**
# PR Description

Please describe the changes proposed in the pull request:

- use existing functions to check type and presence of columns

**This PR improves #704 .**

# PR Checklist

Please ensure that you have done the following:

1. [x] PR in from a fork off your branch. Do not PR from `<your_username>`:`dev`, but rather from `<your_username>`:`<feature-branch_name>`.

2. [x] If you're not on the contributors list, add yourself to `AUTHORS.rst`.

3. [x] Add a line to `CHANGELOG.rst` under the latest version header (i.e. the one that is "on deck") describing the contribution.
    - Do use some discretion here; if there are multiple PRs that are related, keep them in a single line.

## Quick Check

To do a very quick check that everything is correct, follow these steps below:

- [x] Run the command `make check` from pyjanitor's top-level directory. This will automatically run:
    - black formatting
    - flake8 checking
    - running the test suite
    - docs build

Once done, please check off the check-box above.

If `make check` does not work for you, you can execute the commands listed in the Makefile individually.

## Code Changes

If you are adding code changes, please ensure the following:

- [x] Ensure that you have added tests.
- [x] Run all tests (`$ pytest .`) locally on your machine.
    - [x] Check to ensure that test coverage covers the lines of code that you have added.
    - [x] Ensure that all tests pass.

## Documentation Changes

If you are adding documentation changes, please ensure the following:

- [x] Build the docs locally.
- [x] View the docs to check that it renders correctly.

# Relevant Reviewers

Please tag maintainers to review.

- @ericmjl

### CODE REVIEW

**Comment by samukweku** [comment]
@ericmjl @hectormz @sallyhong and others. Need your advice on the API. Currently different values for different columns cannot be applied for the limit parameter; if there are multiple columns, the same `limit` will be applied to all the columns. I am of the opinion that the API should be changed to allow different `limit` for different columns. It looks likely to break current API. Or should I just leave as is; probably users don’t care much for `limit`?

File: CHANGELOG.rst [hunk 1]
```diff
@@ -13,6 +13,7 @@ new version (on deck)
 - [ENH] Add ``pivot_wider`` function, which is the inverse of the `pivot_longer`
   function. @samukweku
 - [INF] Add `openpyxl` to `environment-dev.yml`. @samukweku
+- [ENH] Reduce code by reusing existing functions for fill_direction. @samukweku
```
Comment (ericmjl) [inline CHANGELOG.rst:16]: :tada:

File: janitor/functions.py [hunk 2]
```diff
@@ -4332,56 +4350,51 @@ def fill_direction(
...
+            raise ValueError(
+                """
+                The direction should be a string and should be one of
+                `up`, `down`, `updown`, or `downup`.
+                """
+            )
 
+    # TODO: option to specify limit per column; current implementation
+    # is one `limit` for all the columns. Might need refactoring, or an
+    # API change.
```
Comment (ericmjl) [inline janitor/functions.py:4398]: This is a classic "do we apply this function to one column, or multiple columns" situation! 

As the library matures, I've grown to realize that having singular and plural functions is not necessarily the best, so allowing a function to be applied to multiple columns simultaneously is probably the better way. There must be some way we can "compose" them together, I think, such that we get the desired behaviour. I'm kind of inspired by what I see in JAX, where we have `vmap` to map a function over an array axis. This makes composing functions together more natural.

Maybe a better solution to this problem is to provide a `.map_columns()` function, which allows us to map any pyjanitor function across columns. Let's talk about this in another setting. For now, I like what you've done here, @samukweku, in that you've left this API piece as a TODO to think about later on.

File: tests/functions/test_fill_direction.py [hunk 3]
```diff
@@ -85,35 +85,35 @@ def df():
 
 
 def test_fill_column(df):
-    """Fill down on a single column."""
+    "Fill down on a single column."
```
Comment (ericmjl) [inline tests/functions/test_fill_direction.py]: Docstrings! They should be triple-quoted.

**Review by ericmjl — CHANGES_REQUESTED** [review]
The only change I'm requesting, @samukweku, is about docstring triple quotes, nothing else!

**Review by ericmjl — APPROVED** [review]
I approve, @samukweku! Let's get this one merged in. You've done a massive job with recent PRs, please take the honour of merging this one!

**Pull request merged**